docrawl/docrawl_core.py

```diff
@@ -816,16 +763,65 @@ def extract_table_xpath(browser, page, inp):
     with open(filename, 'wb') as pickle_file:
         pickle.dump(df, pickle_file)
 
-    # data = result
 
-    # print("DATA", data)
-    # with open(filename, "w+", encoding="utf-8") as f:
-    # pass
-    # for i,row in enumerate(data):
-    # print("B",i,row)
-    # f.write(row+"\n")
-    # print("C")
-    # print(data)
+class BrowserMetaData(UserDict):
+    """
+    Stores browser metadata such as driver, proxy, request and function info.
+
+    Ensures synchronisation with .kpv file.
+
+    Expected structure:
+
+    browser_metadata = {
+        "browser": {"driver": "Firefox", "headless": True, "browser_pid": 1234},
+        "proxy": {"ip": "192.123.23.2", "port": 17, "username": "qwerty", "password": "12345"},
+        "request": {"url": "https://forloop.ai", "loaded": True},
+        "function": {"name": "extract_xpath", "input": "/html/main/div[4]/span", "done": False}
+    }
+    """
+
+    def __init__(self, kpv_file: str = 'browser_meta_data.kpv'):
+        super().__init__()
+
+        self.kpv_file = kpv_file
+        self._set_init_values()
+
+    def __str__(self):
+        """
+        Prints metadata in certain order.
+        """
+
+        elements_order = ['browser', 'proxy', 'request', 'function']
+        ordered_dict = dict()
+
+        for elem in elements_order:
+            ordered_dict[elem] = self.data.get(elem)
+
+        return str(ordered_dict)
+
+    def _set_init_values(self):
+        """
+        Initial value for function must be set so crawler is able to start requests.
+        """
+
+        init_function = {"name": "print", "input": "Bla", "done": False}
+        self.__setitem__('function', init_function)
+
+    def __setitem__(self, key, value):
+        VarSafe(value, key, key)
+        save_variables(kept_variables, self.kpv_file)
+        super().__setitem__(key, value)
+
+    def __getitem__(self, key):
+        try:
+            value = load_variable_safe(self.kpv_file, key)
+        except Exception as e:
+            value = None
+
+        # Update value in dict
+        self.__setitem__(key, value)
+
+        return self.data[key]
```

Comment:
not sure if setting these dunder methods is the best approach - I think if there's standard function (method of this class) doing modification of the dict values + manipulating kpv then its little bit easier to read for people who dont know that behaviour of this class has changed

Reply:
@tsai00

Reply:
I first tried to do that with basic getters and setters (see commit 83e9b75), but imho it was more complicated than implementing this class as from UserDict